tensorflow/c/c_api.cc

```diff
@@ -1812,11 +1812,19 @@ void TF_GraphImportGraphDef(TF_Graph* graph, const TF_Buffer* graph_def,
 // While loop functions -------------------------------------------------------
 
 namespace {
-bool CreateInput(const TF_Output& parent_input, TF_Graph* g, const char* name,
-                 TF_Output* input, TF_Status* status) {
+bool CreateInput(TF_Graph* parent_graph, const TF_Output& parent_input, TF_Graph* g,
+                 const char* name, TF_Output* input, TF_Status* status) {
   TF_OperationDescription* desc = TF_NewOperation(g, "Placeholder", name);
   TF_SetAttrType(desc, "dtype", TF_OperationOutputType(parent_input));
-  // TODO(skyewm): set placeholder shape
+  // Set placeholder shape.
+  int num_dims = TF_GraphGetTensorNumDims(parent_graph, parent_input, status);
+  if (status->status.ok() && num_dims >= 0) {
+    int64_t* dims = new int64_t[num_dims];
```

Comment:
Use a vector instead of new (std::vector<int64_t> dims(num_dims), and then use dims.data() and dims.size()). I believe you are leaking dims, and using raw memory allocations like this is in general very dangerous.